# Assignment

In this assignment, you'll be working with the U.S. Education Dataset from Kaggle. The data gives detailed state level information on several facets of the state of education on an annual basis.

To complete this assignment, submit a link to a Jupyter notebook containing your solutions to the following tasks:

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import warnings

warnings.filterwarnings('ignore')

In [2]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

education_df = pd.read_sql_query('select * from useducation',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

#### 1. Determine all the variable types and find the fraction of the missing values for each variable.

In [3]:
# Variable types

education_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1492 entries, 0 to 1491
Data columns (total 25 columns):
PRIMARY_KEY                     1492 non-null object
STATE                           1492 non-null object
YEAR                            1492 non-null int64
ENROLL                          1229 non-null float64
TOTAL_REVENUE                   1280 non-null float64
FEDERAL_REVENUE                 1280 non-null float64
STATE_REVENUE                   1280 non-null float64
LOCAL_REVENUE                   1280 non-null float64
TOTAL_EXPENDITURE               1280 non-null float64
INSTRUCTION_EXPENDITURE         1280 non-null float64
SUPPORT_SERVICES_EXPENDITURE    1280 non-null float64
OTHER_EXPENDITURE               1229 non-null float64
CAPITAL_OUTLAY_EXPENDITURE      1280 non-null float64
GRADES_PK_G                     1319 non-null float64
GRADES_KG_G                     1360 non-null float64
GRADES_4_G                      1361 non-null float64
GRADES_8_G                      1

In [4]:
# Percentage of missing values 

education_df.isnull().sum()*100/education_df.isnull().count()

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          17.627346
TOTAL_REVENUE                   14.209115
FEDERAL_REVENUE                 14.209115
STATE_REVENUE                   14.209115
LOCAL_REVENUE                   14.209115
TOTAL_EXPENDITURE               14.209115
INSTRUCTION_EXPENDITURE         14.209115
SUPPORT_SERVICES_EXPENDITURE    14.209115
OTHER_EXPENDITURE               17.627346
CAPITAL_OUTLAY_EXPENDITURE      14.209115
GRADES_PK_G                     11.595174
GRADES_KG_G                      8.847185
GRADES_4_G                       8.780161
GRADES_8_G                       8.780161
GRADES_12_G                      8.780161
GRADES_1_8_G                     8.780161
GRADES_9_12_G                    8.780161
GRADES_ALL_G                    11.595174
AVG_MATH_4_SCORE                64.075067
AVG_MATH_8_SCORE                64.343164
AVG_READING_4_SCORE             64

#### 2. Notice that the data has a time dimension (year). For this assignment, forget about time and treat all the observations as if they're from the same year. Choose a strategy to deal with the missing values for each variables. For which variables would filling in the missing values with some value make sense? For which might tossing out the records entirely make sense?

In [5]:
# Replace missing values with mean 

education2_df = education_df.copy()

null_val = ['ENROLL','TOTAL_REVENUE','FEDERAL_REVENUE','STATE_REVENUE','LOCAL_REVENUE',
           'TOTAL_EXPENDITURE','INSTRUCTION_EXPENDITURE','SUPPORT_SERVICES_EXPENDITURE',
           'OTHER_EXPENDITURE','CAPITAL_OUTLAY_EXPENDITURE','GRADES_PK_G','GRADES_KG_G',
           'GRADES_4_G','GRADES_8_G','GRADES_12_G','GRADES_1_8_G','GRADES_9_12_G','GRADES_ALL_G',
           'AVG_MATH_4_SCORE','AVG_MATH_8_SCORE','AVG_READING_4_SCORE','AVG_READING_8_SCORE']

for col in null_val:
    education2_df.loc[:, col].fillna(education2_df.loc[:, col].mean(), inplace=True)

In [6]:
# Percentage of missing values 

education2_df.isnull().sum()*100/education2_df.isnull().count()

PRIMARY_KEY                     0.0
STATE                           0.0
YEAR                            0.0
ENROLL                          0.0
TOTAL_REVENUE                   0.0
FEDERAL_REVENUE                 0.0
STATE_REVENUE                   0.0
LOCAL_REVENUE                   0.0
TOTAL_EXPENDITURE               0.0
INSTRUCTION_EXPENDITURE         0.0
SUPPORT_SERVICES_EXPENDITURE    0.0
OTHER_EXPENDITURE               0.0
CAPITAL_OUTLAY_EXPENDITURE      0.0
GRADES_PK_G                     0.0
GRADES_KG_G                     0.0
GRADES_4_G                      0.0
GRADES_8_G                      0.0
GRADES_12_G                     0.0
GRADES_1_8_G                    0.0
GRADES_9_12_G                   0.0
GRADES_ALL_G                    0.0
AVG_MATH_4_SCORE                0.0
AVG_MATH_8_SCORE                0.0
AVG_READING_4_SCORE             0.0
AVG_READING_8_SCORE             0.0
dtype: float64

#### 3. Now, take into account the time factor. Replicate your second answer but this time fill in the missing values by using a statistic that is calculated within the year of the observation. For example, if you want to fill a missing value for a variable with the mean of that variable, calculate the mean by using only the observations for that specific year.

In [7]:
# Replace missing year values with mean 

education3_df = education_df.copy()

years = education3_df["YEAR"].unique()

for col in null_val:
    for year in years:
        education3_df.loc[education3_df["YEAR"] == year, col].fillna(
        education3_df[education3_df["YEAR"] == year][col].mean(), inplace=True)

In [8]:
education3_df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

          ...           GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0         ...              57948.0     58025.0      41167.0      471564.0   
1         ...               9748.0      8789.0       6714.0       79117.0   
2         ...              55433.0     49081.0      37410.0      437127.0   
3         ...              34632.0     36011.0      27651.0      281338.0   
4         ...             418418.0    363296.0     270675.0     3286034.0   

   GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0       196386.0      676174.0        208.327876        252.187522   
1        30847.0      112335.0               NaN               NaN   
2       175210.0      614881.0        215.253932        265.366278   
3       123113.0      405259.0        210.206028        256.312090   
4      1372011.0     4717112.0        208.398961        260.892247   

   AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0           207.963517                  NaN  
1                  NaN           258.859712  
2           206.212716           262.169895  
3           208.634458           264.619665  
4           196.764414                  NaN  

[5 rows x 25 columns]

#### 4. This time, fill in the missing values using interpolation (extrapolation).

In [9]:
# Replace missing values with interpolation 

education4_df = education_df.copy()

for col in null_val:
    education4_df.loc[:, col].interpolate(inplace=True)

In [10]:
education4_df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

          ...           GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0         ...              57948.0     58025.0      41167.0      471564.0   
1         ...               9748.0      8789.0       6714.0       79117.0   
2         ...              55433.0     49081.0      37410.0      437127.0   
3         ...              34632.0     36011.0      27651.0      281338.0   
4         ...             418418.0    363296.0     270675.0     3286034.0   

   GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0       196386.0      676174.0        208.327876        252.187522   
1        30847.0      112335.0        211.790904        258.776900   
2       175210.0      614881.0        215.253932        265.366278   
3       123113.0      405259.0        210.206028        256.312090   
4      1372011.0     4717112.0        208.398961        260.892247   

   AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0           207.963517                  NaN  
1           207.088116           258.859712  
2           206.212716           262.169895  
3           208.634458           264.619665  
4           196.764414           265.519676  

[5 rows x 25 columns]

#### 5. Compare your results for the 2nd, 3rd, and 4th questions. Do you find any meaningful differences?

In [11]:
# Print Statistics of each column 

for col in null_val:
    print("Statistics for columns: {}".format(col))
    print(pd.concat([education_df[col], education2_df[col],
                     education3_df[col], education4_df[col]], axis=1).describe())

Statistics for columns: ENROLL
             ENROLL        ENROLL        ENROLL        ENROLL
count  1.229000e+03  1.492000e+03  1.229000e+03  1.456000e+03
mean   9.159308e+05  9.159308e+05  9.159308e+05  1.005984e+06
std    1.065280e+06  9.667725e+05  1.065280e+06  1.009158e+06
min    4.386600e+04  4.386600e+04  4.386600e+04  4.386600e+04
25%    2.583140e+05  3.150940e+05  2.583140e+05  2.973995e+05
50%    6.483130e+05  8.204140e+05  6.483130e+05  7.540845e+05
75%    1.014528e+06  9.211780e+05  1.014528e+06  1.595024e+06
max    6.307022e+06  6.307022e+06  6.307022e+06  6.307022e+06
Statistics for columns: TOTAL_REVENUE
       TOTAL_REVENUE  TOTAL_REVENUE  TOTAL_REVENUE  TOTAL_REVENUE
count   1.280000e+03   1.492000e+03   1.280000e+03   1.492000e+03
mean    9.092082e+06   9.092082e+06   9.092082e+06   1.111164e+07
std     1.174519e+07   1.087818e+07   1.174519e+07   1.200155e+07
min     4.656500e+05   4.656500e+05   4.656500e+05   4.656500e+05
25%     2.186305e+06   2.546261e+06   2.186

       AVG_READING_4_SCORE  AVG_READING_4_SCORE  AVG_READING_4_SCORE  \
count           533.000000          1492.000000           533.000000   
mean            218.866154           218.866154           218.866154   
std               7.769616             4.641053             7.769616   
min             178.557612           178.557612           178.557612   
25%             214.663401           218.866154           214.663401   
50%             220.416034           218.866154           220.416034   
75%             223.999337           218.866154           223.999337   
max             236.773867           236.773867           236.773867   

       AVG_READING_4_SCORE  
count          1492.000000  
mean            218.971553  
std               5.959927  
min             178.557612  
25%             216.126818  
50%             219.771325  
75%             222.684611  
max             236.773867  
Statistics for columns: AVG_READING_8_SCORE
       AVG_READING_8_SCORE  AVG_READING_8_SCOR

I find that using the mean to replace missing values reduces the differences between the original and the new mean. In some cases, it won't affect other central tendencies by much either. On the other hand, filling missing values by interpolation can lead to varying differences. For example, the state revenue mean slightly increased by 0.70e+06 after interpolation, but the new central tendencies were relatively close to the original dataset. In other cases, interpolation can change the meaning drastically like in the column total revenue, where the mean decreased from 9.09 e+06 to 1.11e+07.  